File: go/strictcli/routing.go
package strictcli

import (
	"strings"
)

// routeResult holds the outcome of command routing through the group/command tree.
type routeResult struct {
	// Exactly one of cmd, lastGroup, or err will be meaningful.
	cmd       *Command // resolved command, nil if routing stopped at a group or errored
	lastGroup *Group   // deepest group reached (set even when cmd is found)
	path      []string // group names traversed to reach the result
	rest      []string // remaining tokens after routing consumed path + command name

	// Error cases
	err string // non-empty if routing failed (deprecated, unknown, no command)
	// commandPrefix is set for error messages that need the full path context
	commandPrefix string
	// helpAtGroup is true when the user requested help at a group level
	helpAtGroup bool
}

// resolveCommand traverses the group/command/deprecated tree using tokens from
// rest, returning the resolved command or an error. It does not perform command
// parsing or help formatting — callers use the returned routeResult to decide
// the next step.
func (a *App) resolveCommand(rest []string) routeResult {
	currentGroups := a.groups
	currentCommands := a.commands
	currentDeprecated := a.deprecatedMap
	var path []string
	var lastGroup *Group

	for len(rest) > 0 {
		token := rest[0]

		// Check groups first
		if grp, ok := currentGroups[token]; ok {
			path = append(path, token)
			lastGroup = grp
			rest = rest[1:]

			// Check for help at this level
			if len(rest) == 0 || (len(rest) == 1 && (rest[0] == "--help" || rest[0] == "-h")) {
				return routeResult{lastGroup: lastGroup, path: path, rest: rest, helpAtGroup: true}
			}

			// Descend into subgroup
			currentGroups = grp.Groups
			currentCommands = grp.Commands
			currentDeprecated = grp.deprecatedMap
			continue
		}

		// Check commands
		if cmd, ok := currentCommands[token]; ok {
			return routeResult{cmd: cmd, lastGroup: lastGroup, path: path, rest: rest[1:]}
		}

		// Check deprecated commands
		if msg, ok := currentDeprecated[token]; ok {
			return routeResult{
				err:  errCommandDeprecated(token, msg),
				path: path, rest: rest,
			}
		}

		// Unknown command — include path in error message
		if len(path) > 0 {
			prefix := strings.Join(append([]string{a.Name}, path...), " ")
			return routeResult{
				err:           errUnknownCommandInGroup(token, strings.Join(path, " ")),
				commandPrefix: prefix,
				path:          path, rest: rest,
			}
		}
		return routeResult{
			err:  errUnknownCommand(token),
			path: path, rest: rest,
		}
	}

	// Loop ended without finding a command — exhausted by group traversal
	if lastGroup != nil {
		return routeResult{lastGroup: lastGroup, path: path, rest: rest, helpAtGroup: true}
	}
	return routeResult{err: errNoCommandSpecified, path: path, rest: rest}
}

File: typescript/src/routing.ts
/**
 * Command routing: first non-flag token selects a command or group; groups
 * are traversed to arbitrary depth. Mirrors Go resolveCommand (routing.go)
 * with Python _resolve_command as the divergence ground truth: after a group
 * token is consumed, a leading --help/-h in the remaining tokens requests
 * group help regardless of what follows (Python checks segments[0]; Go
 * additionally requires it to be the only remaining token).
 */

import type { AppImpl, GroupImpl, RegisteredCommand } from "./app.js";
import {
	errCommandDeprecated,
	errNoCommandSpecified,
	errUnknownCommand,
	errUnknownCommandInGroup,
} from "./errors.js";

export interface RouteResult {
	/** Resolved command; undefined if routing stopped at a group or errored. */
	readonly cmd?: RegisteredCommand;
	/** Deepest group reached (set even when cmd is found under a group). */
	readonly lastGroup?: GroupImpl;
	/** Group names traversed to reach the result. */
	readonly path: readonly string[];
	/** Remaining tokens after routing consumed path + command name. */
	readonly rest: readonly string[];
	/** Non-empty if routing failed (deprecated, unknown, no command). */
	readonly err?: string;
	/** Full "app group..." prefix for error messages that need path context. */
	readonly commandPrefix?: string;
	/** True when the user requested help at a group level. */
	readonly helpAtGroup: boolean;
}

export function resolveCommand(
	app: AppImpl,
	segments: readonly string[],
): RouteResult {
	let currentGroups: ReadonlyMap<string, GroupImpl> = app.groups;
	let currentCommands: ReadonlyMap<string, RegisteredCommand> = app.commands;
	let currentDeprecated: ReadonlyMap<string, string> = app.deprecated;
	const path: string[] = [];
	let lastGroup: GroupImpl | undefined;
	let rest = segments;

	while (rest.length > 0) {
		const token = rest[0] as string;

		const grp = currentGroups.get(token);
		if (grp !== undefined) {
			path.push(token);
			lastGroup = grp;
			rest = rest.slice(1);

			if (rest.length === 0 || rest[0] === "--help" || rest[0] === "-h") {
				return { lastGroup, path, rest, helpAtGroup: true };
			}

			currentGroups = grp.groups;
			currentCommands = grp.commands;
			currentDeprecated = grp.deprecated;
			continue;
		}

		const cmd = currentCommands.get(token);
		if (cmd !== undefined) {
			return {
				cmd,
				...(lastGroup !== undefined ? { lastGroup } : {}),
				path,
				rest: rest.slice(1),
				helpAtGroup: false,
			};
		}

		const depMsg = currentDeprecated.get(token);
		if (depMsg !== undefined) {
			return {
				err: errCommandDeprecated(token, depMsg),
				path,
				rest,
				helpAtGroup: false,
			};
		}

		if (path.length > 0) {
			return {
				err: errUnknownCommandInGroup(token, path.join(" ")),
				commandPrefix: [app.name, ...path].join(" "),
				path,
				rest,
				helpAtGroup: false,
			};
		}
		return { err: errUnknownCommand(token), path, rest, helpAtGroup: false };
	}

	// Loop ended without finding a command. Callers never pass empty segments
	// and group exhaustion is caught inside the loop, so this is defensive.
	if (lastGroup !== undefined) {
		return { lastGroup, path, rest, helpAtGroup: true };
	}
	return { err: errNoCommandSpecified(), path, rest, helpAtGroup: false };
}
